### DISCUSSION TITLE
need help creating a multi-tap `ctrl+shift` then `alt+shift`

### DISCUSSION THREAD

**Discussion by txtyash** [kmonad/kmonad#573] [Q&A] [answered]
If the key is pressed once then it should act like `C-S` waiting for the next input say `^` then the input should be `ctrl+shift+^`

If the key is pressed twice within a timeout then it should act like `A-S` waiting for the next input say `@` and the input should be `alt+shift+@`

One of my many failed attempts:
```kbd
csft (around
     (sticky-key 500 lsft)
     (sticky-key 500 lctl))
asft (around
     (sticky-key 500 lsft)
     (sticky-key 500 lalt))
casft (multi-tap 300
(around-next @csft)
(around-next @asft))
```

**Comment by MuhammedZakir** [comment]
```clojure
ctsf (around-next (around lctl lsft))
alsf (around-next (around lalt lsft))
macs (multi-tap 300 @ctsf @alsf)
```

This doesn't have the `500`ms timeout for modifers-combo. For that, `sticky-key` is the best way as you have tried. However, it misbehaves when inside `multi-tap`. To be precise, second button in `multi-tap` works correctly, but the first one misbehaves. If you are using `master` branch, another way is (I haven't tested this):
```clojure
sfpo (press-only lsft)
sfro (release-only lsft)
csft #((press-only lctl) @sfpo P500 @sfro (release-only lctl))
asft #((press-only lalt) @sfpo P500 @sfro (release-only lalt))
casft (multi-tap 300 @csft @asft)
```

---

Note to myself/devs: add a `timeout` option for `around-next`.

**Reply by txtyash** [reply]
@MuhammedZakir the first solution doesn't work. The second key works but the first one doesn't. Similar thing happened to one of my bindings.

Currently what I've done is created a layer for `Ctrl+shift` the same key in that layer is `Alt+shift` sticky key.
This works fine but it doesn't allow me to have combinations like `Ctrl+shift+^` as I can only fit alphabets at a time in a single layer.

**Reply by MuhammedZakir** [reply]
It worked for me. I am using `keycode-refactor` branch (that's why, I couldn't test the second solution). Anyway, I will build `master` branch and test both and will report here.

P.S. What about the second one? Did it work?

**Reply by MuhammedZakir** [reply]
I just tested using latest build of `master` branch and the first one worked for me. Below script is a variant of it and it too worked. Can you test them again after building from `master` branch? Don't forget to pull latest changes before building.

```clojure
  csft #((around-next lctl) (around-next lsft))
  asft #((around-next lalt) (around-next lsft))
  mac (multi-tap 300 @csft @asft)
```

**Comment by txtyash** [comment]
@MuhammedZakir I updated the `kmonad-git` package from AUR. But that didn't work. Should I build it myself with stack?

[My config file](https://pastebin.com/f8YBdT1v)

**Reply by MuhammedZakir** [reply]
That package looks fine. Is this your KMonad version -- `0.4.1, commit 3011d359db5b6305625b6f4110457dbec0dc8409`? This should be what you get when building from `master`.

Can you post a debug output of when you use `@mac` button -- `kmonad -ldebug <your config>`?

**Comment by txtyash** [comment]
@MuhammedZakir 
yes the same version.
```console
❯ : kmonad -ldebug ~/.kmonad.kbd
Opening KeySink
Registering Uinput device
Opening KeySource
Initiating ioctl grab
Closing KeySink
Unregistering Uinput device
Closing Uinput device file
kmonad: Could not perform IOCTL grab on: /dev/input/by-path/platform-i8042-serio-0-event-kbd
```

**Reply by MuhammedZakir** [reply]
When you tested the config, you killed currently running `kmonad` and started it again with the new config, right? You must do that when you make changes.

**Reply by MuhammedZakir** [reply]
With this config:
```clojure
  csft #((around-next lctl) (around-next lsft))
  asft #((around-next lalt) (around-next lsft))
  mac (multi-tap 300 @csft @asft)
```

with `v` mapped to `@mac`, this is the debug output I get for `Tv Tt`:

```
Got: Press v
Running hooks
Registering untimed hook: 75
Registering Ms {unMs = 300}ms hook: 74
Got: Release v
Running hooks
Registering Ms {unMs = 196}ms hook: 76
Cancelling hook: 76
Registering untimed hook: 77
Registering untimed hook: 78
Registering untimed hook: 79
Tried cancelling expired hook: 74
Got: Press t
Running hooks
Registering untimed hook: 80
Putting: Press lctl
Registering untimed hook: 81
Putting: Press lsft
Registering untimed hook: 82
Registering untimed hook: 83
Putting: Press t
Got: Release t
Running hooks
Registering untimed hook: 84
Putting: Release lctl
Putting: Release lsft
Putting: Release t
```

**Reply by txtyash** [reply]
> When you tested the config, you killed currently running `kmonad` and started it again with the new config, right? You must do that when you make changes.

yes I restart kmonad's service I made with:
`systemctl restart kmon`

Your binding works. Until now I'd mapped that to a sophisticated key which I think should work but don't know why it isn't. Your binding worked when I mapped it to a key on my first layer, while it didn't completely work when I mapped it in a layer called `rshift` which is toggled when I hold the quote(') key. Only `alt+shift` worked in that layer.

My config file https://pastebin.com/f8YBdT1v

**Reply by MuhammedZakir** [reply]
It had to do with timing issues. If you press `'` for a little longer than `space`, it wil work.

This is due to current implementation; it is very fragile, especially the composibility of buttons.This is why @david-janssen is doing (almost) complete refactoring of internals.

**Comment by txtyash** [comment, marked as answer]
@MuhammedZakir Okay, It works then. I'll keep an eye on future updates. Thanks for all the help.
